### PR TITLE
Add conditional compilation symbol FEATURE_REMOTING

### PR DESCRIPTION
**Pull request by alinapopa**

### CODE REVIEW

File: src/Castle.Windsor/MicroKernel/Lifestyle/Scoped/CallContextLifetimeScope.cs [hunk 1]
```diff
@@ -38,7 +42,11 @@ public class CallContextLifetimeScope : ILifetimeScope
 		private static readonly ConcurrentDictionary<Guid, CallContextLifetimeScope> appDomainLocalInstanceCache =
 			new ConcurrentDictionary<Guid, CallContextLifetimeScope>();
 
+#if FEATURE_REMOTING
 		private static readonly string keyInCallContext = "castle.lifetime-scope-" + AppDomain.CurrentDomain.Id.ToString(CultureInfo.InvariantCulture);
+#else
+		private static readonly AsyncLocal<Guid> CallContextData = new AsyncLocal<Guid>();
```
Comment (jonorossi) [inline src/Castle.Windsor/MicroKernel/Lifestyle/Scoped/CallContextLifetimeScope.cs]: Would have been nice if `AsyncLocal<T>` was available in .NET 4.5.

File: src/Castle.Windsor/Windsor/WindsorContainer.cs [hunk 2]
```diff
@@ -36,11 +36,11 @@ namespace Castle.Windsor
 	/// </summary>
 	[Serializable]
 	[DebuggerDisplay("{name,nq}")]
-#if (SILVERLIGHT)
-	public partial class WindsorContainer : IWindsorContainer
-#else
+#if FEATURE_REMOTING
 	[DebuggerTypeProxy(typeof(KernelDebuggerProxy))]
```
Comment (jonorossi) [inline src/Castle.Windsor/Windsor/WindsorContainer.cs]: It looks like `DebuggerTypeProxy` is supported on .NET Core.

File: src/Castle.Windsor/Windsor/WindsorContainer.cs [hunk 3]
```diff
@@ -36,11 +36,11 @@ namespace Castle.Windsor
...
 	public partial class WindsorContainer : MarshalByRefObject, IWindsorContainer
```
Comment (jonorossi) [inline src/Castle.Windsor/Windsor/WindsorContainer.cs]: Could you put the `#if FEATURE_REMOTING` around just `MarshalByRefObject` like many other classes (i.e. add new lines) so the two definitions can't get out of sync.

File: src/Castle.Windsor/MicroKernel/DefaultKernel.cs [hunk 4]
```diff
@@ -46,7 +46,7 @@ namespace Castle.MicroKernel
 	///   Default implementation of <see cref = "IKernel" />. This implementation is complete and also support a kernel hierarchy (sub containers).
 	/// </summary>
 	[Serializable]
-#if !SILVERLIGHT
+#if FEATURE_REMOTING
 	[DebuggerTypeProxy(typeof(KernelDebuggerProxy))]
```
Comment (jonorossi) [inline src/Castle.Windsor/MicroKernel/DefaultKernel.cs]: DebuggerTypeProxy again.

File: src/Castle.Windsor/MicroKernel/DefaultKernel.cs [hunk 5]
```diff
@@ -46,7 +46,7 @@ namespace Castle.MicroKernel
...
 	public partial class DefaultKernel : MarshalByRefObject, IKernel, IKernelEvents, IKernelInternal
```
Comment (jonorossi) [inline src/Castle.Windsor/MicroKernel/DefaultKernel.cs]: New lines around MarshalByRefObject again.

File: src/Castle.Windsor/MicroKernel/Lifestyle/Scoped/CallContextLifetimeScope.cs (hunk 1 above)
Comment (alinapopa) [inline src/Castle.Windsor/MicroKernel/Lifestyle/Scoped/CallContextLifetimeScope.cs]: I have added FEATURE_ASYNCLOCAL to be used with compilation for netstandard.

File: src/Castle.Windsor/Windsor/WindsorContainer.cs (hunk 3 above)
Comment (alinapopa) [inline src/Castle.Windsor/Windsor/WindsorContainer.cs]: Done

File: src/Castle.Windsor/MicroKernel/DefaultKernel.cs (hunk 5 above)
Comment (alinapopa) [inline src/Castle.Windsor/MicroKernel/DefaultKernel.cs]: Done

File: src/Castle.Windsor/MicroKernel/Lifestyle/Scoped/ThreadStaticLifetimeScope.cs [hunk 6]
```diff
@@ -12,9 +12,9 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+#if !FEATURE_REMOTING && !FEATURE_ASYNCLOCAL
 namespace Castle.MicroKernel.Lifestyle.Scoped
 {
-#if SILVERLIGHT
```
Comment (jonorossi) [inline src/Castle.Windsor/MicroKernel/Lifestyle/Scoped/ThreadStaticLifetimeScope.cs]: I just realised we can actually just delete `ThreadStaticLifetimeScope` since it was only used on Silverlight, it'll also simplify some of the preprocessor conditionals in other files.

File: src/Castle.Windsor/MicroKernel/Lifestyle/Scoped/CallContextLifetimeScope.cs (hunk 1 above)
Comment (jonorossi) [inline src/Castle.Windsor/MicroKernel/Lifestyle/Scoped/CallContextLifetimeScope.cs]: Oh, I was just making an off handed remark about it not existing until .NET 4.6, but that's fine to have both. We'll need to get our README updated soon with descriptions of these defines like Castle Core, can be another PR.

File: src/Castle.Windsor/MicroKernel/Lifestyle/Scoped/ThreadStaticLifetimeScope.cs (hunk 6 above)
Comment (alinapopa) [inline src/Castle.Windsor/MicroKernel/Lifestyle/Scoped/ThreadStaticLifetimeScope.cs]: In this case I would rather remove FEATURE_ASYNCLOCAL because there will be needed  `#if FEATURE_REMOTING || FEATURE_ASYNCLOCAL` in lots of places

Comment (jonorossi) [inline src/Castle.Windsor/MicroKernel/Lifestyle/Scoped/ThreadStaticLifetimeScope.cs]: Go for it. My view is that we want conditionals to add features back in for old stuff (not yet or never going to be supported on the new runtime) rather than conditionals for what should be supported going forward.

Comment (alinapopa) [inline src/Castle.Windsor/MicroKernel/Lifestyle/Scoped/ThreadStaticLifetimeScope.cs]: Removed `ThreadStaticLifetimeScope` and `FEATURE_ASYNCLOCAL`
